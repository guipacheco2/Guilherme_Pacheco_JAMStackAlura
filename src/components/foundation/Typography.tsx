import React from 'react'
import styled, { css, CSSProperties } from 'styled-components'
import {
  breakpointsMedia,
  ColorKeys,
  createBreakpoints,
  propsToStyle,
  ResponsiveBreakpoints,
  TypographyVariantKeys,
} from '../../theme'
import { Link } from '../commons'

export interface StyledTypographyProps {
  htmlFor?: string
  onColor: ColorKeys
  textAlign?: ResponsiveBreakpoints<CSSProperties['textAlign']>
  variant:
    | TypographyVariantKeys
    | { md: TypographyVariantKeys; xs: TypographyVariantKeys }
}

const StyledTypography = styled.span<StyledTypographyProps>(
  ({ onColor, textAlign, theme, variant }) => {
    return css`
      ${() => {
        if (typeof variant === 'string') {
          return theme.typographyVariants[variant]
        }

        return breakpointsMedia(
          createBreakpoints(variant, (v) => theme.typographyVariants[v]),
        )
      }}
      color: ${theme.colors[onColor].contrastText};
      ${propsToStyle({ textAlign })}
    `
  },
)

interface TypographyProps<C extends React.ElementType = React.ElementType>
  extends StyledTypographyProps {
  as?: C
  children?: React.ReactNode
  href?: string
}

export type TypographyPropsGeneric<
  C extends React.ElementType
> = TypographyProps<C> & Omit<React.ComponentProps<C>, keyof TypographyProps>

export function Typography<C extends React.ElementType = 'span'>({
  as,
  children,
  href,
  htmlFor,
  onColor,
  textAlign,
  variant,
}: TypographyPropsGeneric<C>): JSX.Element {
  if (href) {
    return (
      <Link href={href}>
        <StyledTypography
          as={as as never}
          variant={variant}
          htmlFor={htmlFor}
          onColor={onColor}
          textAlign={textAlign}
        >
          {children}
        </StyledTypography>
      </Link>
    )
  }

  return (
    <StyledTypography
      as={as as never}
      variant={variant}
      htmlFor={htmlFor}
      onColor={onColor}
      textAlign={textAlign}
    >
      {children}
    </StyledTypography>
  )
}
