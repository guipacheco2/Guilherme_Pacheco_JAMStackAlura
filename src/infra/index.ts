export * from './hooks'
export * from './test'
