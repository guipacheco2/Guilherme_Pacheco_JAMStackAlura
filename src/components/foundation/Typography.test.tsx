import React from 'react'
import { render } from '../../infra'
import { Typography } from './Typography'

describe('<Typography />', () => {
  describe('when variant', () => {
    test('is simple', () => {
      const { baseElement } = render(
        <Typography surfaceColor="primary" variant="bodyText1">
          child
        </Typography>,
      )

      expect(baseElement).toMatchSnapshot()
    })

    test('is responsive', () => {
      const { baseElement } = render(
        <Typography
          surfaceColor="primary"
          variant={{ md: 'bodyText2', xs: 'bodyText1' }}
        >
          child
        </Typography>,
      )

      expect(baseElement).toMatchSnapshot()
    })
  })

  describe('when href', () => {
    test('is present', () => {
      const { baseElement } = render(
        <Typography surfaceColor="primary" variant="bodyText1" href="/xpto">
          child
        </Typography>,
      )

      expect(baseElement).toMatchSnapshot()
    })
  })
})
