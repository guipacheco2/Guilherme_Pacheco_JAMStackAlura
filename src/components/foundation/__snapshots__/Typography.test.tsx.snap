// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Typography /> when href is present 1`] = `
<body>
  .c0 {
  -webkit-text-decoration: none;
  text-decoration: none;
  opacity: 1;
}

.c0:hover,
.c0:focus {
  opacity: 0.5;
}

.c1 {
  font-size: 16px;
  font-weight: 400;
  -webkit-letter-spacing: 0.5px;
  -moz-letter-spacing: 0.5px;
  -ms-letter-spacing: 0.5px;
  letter-spacing: 0.5px;
  color: #FAFAFA;
}

<div>
    <a
      class="c0"
      href="/xpto"
    >
      <span
        class="c1"
      >
        child
      </span>
    </a>
  </div>
</body>
`;

exports[`<Typography /> when variant is responsive 1`] = `
<body>
  .c0 {
  color: #FAFAFA;
}

@media only screen and (min-width:0px) {
  .c0 {
    font-size: 16px;
    font-weight: 400;
    -webkit-letter-spacing: 0.5px;
    -moz-letter-spacing: 0.5px;
    -ms-letter-spacing: 0.5px;
    letter-spacing: 0.5px;
  }
}

@media only screen and (min-width:1024px) {
  .c0 {
    font-size: 14px;
    font-weight: 400;
    -webkit-letter-spacing: 0.25px;
    -moz-letter-spacing: 0.25px;
    -ms-letter-spacing: 0.25px;
    letter-spacing: 0.25px;
  }
}

<div>
    <span
      class="c0"
    >
      child
    </span>
  </div>
</body>
`;

exports[`<Typography /> when variant is simple 1`] = `
.c0 {
  font-size: 16px;
  font-weight: 400;
  -webkit-letter-spacing: 0.5px;
  -moz-letter-spacing: 0.5px;
  -ms-letter-spacing: 0.5px;
  letter-spacing: 0.5px;
  color: #FAFAFA;
}

<body>
  <div>
    <span
      class="c0"
    >
      child
    </span>
  </div>
</body>
`;
